### PR TITLE
Refactoring of the FSM's latest operation

### PR DESCRIPTION
**Pull request by velzevur**
This is a preparatory PR before [SC FSM: handling of on-chain pinned env and forks awareness](https://www.pivotaltracker.com/story/show/165214367)

Not user visible, hence no line in release notes

### CODE REVIEW

File: apps/aechannel/src/aesc_fsm.erl [hunk 1]
```diff
@@ -618,81 +638,100 @@ awaiting_signature(cast, {?SIGNED, ?FND_CREATED, SignedTx} = Msg,
...
+                   #data{op = #op_sign{tag = ?SHUTDOWN
+                                     , data = OpData0}} = D) ->
+    #op_data{updates = Updates} = OpData0,
     maybe_check_sigs(SignedTx, Updates, aesc_close_mutual_tx, not_close_mutual_tx, me,
         fun() ->
             D1 = send_shutdown_msg(SignedTx, D),
-            D2 = D1#data{latest = {shutdown, SignedTx, Updates},
+            OpData = OpData0#op_data{signed_tx = SignedTx},
+            D2 = D1#data{op = #op_ack{tag = shutdown
+                                    , data = OpData},
```
Comment (uwiger) [inline apps/aechannel/src/aesc_fsm.erl]: move comma to line below? Also, aren't we doing `{ tag ...` and aligning the comma with the `{`?

File: apps/aechannel/src/aesc_fsm.erl [hunk 2]
```diff
@@ -999,12 +1054,14 @@ wdraw_half_signed(cast, {Req,_} = Msg, D) when ?UPDATE_CAST(Req) ->
     %% arrived.
     lager:debug("race detected: ~p", [Msg]),
     handle_update_conflict(Req, D);
-wdraw_half_signed(cast, {?WDRAW_SIGNED, Msg}, #data{latest = Latest} = D) ->
+wdraw_half_signed(cast, {?WDRAW_SIGNED, Msg}
+                  , #data{op = #op_ack{tag = withdraw_tx} = Op} = D) ->
```
Comment (uwiger) [inline apps/aechannel/src/aesc_fsm.erl]: I don't know ... it feels more natural to have that comma on the preceding line.

File: apps/aechannel/src/aesc_fsm.erl [hunk 3]
```diff
@@ -1031,12 +1088,15 @@ wdraw_half_signed(Type, Msg, D) ->
 %% wait for confirmation; postpone instead.
 %%
 wdraw_signed(enter, _OldSt, _D) -> keep_state_and_data;
-wdraw_signed(cast, {?WDRAW_LOCKED, Msg}, #data{latest = {withdraw, SignedTx,
-                                                         Updates}} = D) ->
+wdraw_signed(cast, {?WDRAW_LOCKED, Msg},
```
Comment (uwiger) [inline apps/aechannel/src/aesc_fsm.erl]: (See comment above) - like this

File: apps/aechannel/src/aesc_fsm.erl [hunk 4]
```diff
@@ -1625,12 +1687,14 @@ new_contract_tx_for_signing(Opts, From, #data{ state = State
                                                     Deposit, CallData)],
     {OnChainEnv, OnChainTrees} = aetx_env:tx_env_and_trees_from_top(aetx_contract),
     Height = aetx_env:height(OnChainEnv),
+    %% TODO PT-165214367: maybe set block_hash
+    BlockHash = ?NOT_SET_BLOCK_HASH,
     ActiveProtocol = aec_hard_forks:protocol_effective_at_height(Height),
-    try
+    try
```
Comment (uwiger) [inline apps/aechannel/src/aesc_fsm.erl]: trailing space

File: apps/aechannel/src/aesc_fsm.erl [hunk 5]
```diff
@@ -1759,24 +1827,28 @@ check_deposit_created_msg(#{ channel_id := ChanId
...
 
 check_deposit_signed_msg(#{ channel_id := ChanId
-                          , block_hash := ?NOT_SET_BLOCK_HASH
-                          , data       := #{tx := TxBin}} = Msg,
-                          #data{ on_chain_id = ChanId
-                               , latest = {ack, deposit_tx, _, Updates} } = Data) ->
+                          , block_hash := _BlockHash
+                          , data       := #{tx := TxBin}} = Msg
+                          , #data{ on_chain_id = ChanId
+                                , op = #op_ack{tag = deposit_tx
```
Comment (uwiger) [inline apps/aechannel/src/aesc_fsm.erl]: Indent one more space

File: apps/aechannel/src/aesc_fsm.erl [hunk 6]
```diff
@@ -2017,9 +2093,9 @@ check_update_ack_msg_(#{ channel_id := ChanId
     end.
 
 check_signed_update_ack_tx(SignedTx, Msg,
-                           #data{state = State, opts = Opts,
-                                latest = Latest} = D) ->
-    {ack, ?UPDATE, _, Updates} = Latest,
+                           #data{state = State, opts = Opts
+                               , op = #op_ack{tag = ?UPDATE} = Op} = D) ->
```
Comment (uwiger) [inline apps/aechannel/src/aesc_fsm.erl]: Indentation, suggest:
```
    #data{ state = ...
         , op = ...
```

File: apps/aechannel/src/aesc_fsm.erl [hunk 7]
```diff
@@ -2114,14 +2192,15 @@ send_shutdown_ack_msg(SignedTx, #data{session = Session} = Data) ->
...
      , block_hash => ?NOT_SET_BLOCK_HASH
      , data       => #{tx => TxBin} }.
 
 check_shutdown_msg(_Msg, #data{ client_connected = false }) ->
     {error, client_disconnected};
-check_shutdown_msg(#{channel_id := ChanId,
-                     block_hash := ?NOT_SET_BLOCK_HASH,
-                     data := #{tx := TxBin}} = Msg,
+check_shutdown_msg(#{channel_id := ChanId
+                   , block_hash := BlockHash
```
Comment (uwiger) [inline apps/aechannel/src/aesc_fsm.erl]: Indentation, see above

**Comment by velzevur** [comment]
```

%%% aecore_errorfree_SUITE ==> application_test: FAILED
%%% aecore_errorfree_SUITE ==> {{badmatch,[{dev1,{"aeternity.log",
                   <<"2019-08-05 13:21:07.911 [error] emulator Error in process <0.663.0> on node aeternity_dev1@localhost with exit value:">>}},
            {dev1,{"aeternity_mining.log",
                   <<"2019-08-05 13:21:07.911 [error] <0.442.0>@aec_conductor:handle_monitor_message:434 Worker died: {mining,<0.663.0>,{badarg,[{erlang,list_to_integer,[\"7dbfVerified\",16],[]},{aeminer_pow_cuckoo,'-parse_generation_result/2-lc$^0/1-0-',1,[{file,\"/home/builder/aeternity/_build/default/lib/aeminer/src/aeminer_pow_cuckoo.erl\"},{line,496}]},{aeminer_pow_cuckoo,'-parse_generation_result/2-lc$^0/1-0-',1,[{file,\"/home/builder/aeternity/_build/default/lib/aeminer/src/aeminer_pow_cuckoo.erl\"},{line,496}]},{aeminer_pow_cuckoo,parse_generation_result,2,[{file,\"/home/builder/aeternity/_build/default/lib/aeminer/src/aeminer_pow_cuckoo.erl\"},{line,496}]},{aeminer_pow_cuckoo,generate_int,7,[{file,\"/home/builder/aeternity/_build/default/lib/aeminer/src/aeminer_pow_cuckoo.erl\"},{line,273}]},{aeminer_pow_cuckoo,generate_from_hash,5,[{file,\"/home/builder/aeternity/_build/default/lib/aeminer/src/aeminer_pow_cuckoo.erl\"},{line,195}]},{aec_conductor,'-start_mining_/1-fun-4-',5,[{file,\"/home/builder/aeternity/apps/aecore/src/aec_conductor.erl\"},{line,826}]},{aec_conductor,'-wrap_worker_fun/1-fun-0-',2,[{file,\"/home/builder/aeternity/apps/aecore/src/aec_conductor.erl\"},{line,543}]}]}}">>}}]},
 [{aecore_errorfree_SUITE,application_test,1,
                          [{file,"/home/builder/aeternity/apps/aecore/test/aecore_errorfree_SUITE.erl"},
                           {line,60}]},
  {test_server,ts_tc,3,[{file,"test_server.erl"},{line,1546}]},
  {test_server,run_test_case_eval1,6,[{file,"test_server.erl"},{line,1062}]},
  {test_server,run_test_case_eval,9,[{file,"test_server.erl"},{line,994}]}]}
```

File: apps/aechannel/src/aesc_fsm.hrl [hunk 8]
```diff
@@ -193,21 +196,46 @@
...
-                           | close,
-                      TxHash :: binary(), aetx_sign:signed_tx(),
-                      [aesc_offchain_update:update()]}
-                   | {reestablish, OffChainTx :: aetx_sign:signed_tx()}.
+-record(op_data, {signed_tx  :: aetx_sign:signed_tx(),
+                  updates    :: [aesc_offchain_update:update()],
+                  block_hash :: aec_blocks:block_header_hash()
+                  }).
+
+-record(op_sign, { tag :: sign_tag()
```
Comment (grzegorz225) [inline apps/aechannel/src/aesc_fsm.hrl]: Add spaces to align ``::``

File: apps/aechannel/src/aesc_fsm.erl [hunk 9]
```diff
@@ -618,81 +634,100 @@ awaiting_signature(cast, {?SIGNED, ?FND_CREATED, SignedTx} = Msg,
...
             D1 = send_update_ack_msg(SignedTx, D),
             {OnChainEnv, OnChainTrees} =
                 aetx_env:tx_env_and_trees_from_top(aetx_contract),
             State = aesc_offchain_state:set_signed_tx(SignedTx, Updates, D1#data.state,
                                                       OnChainTrees, OnChainEnv, Opts),
-            D2 = D1#data{log   = log_msg(rcv, ?SIGNED, Msg, D1#data.log),
-                        state = State,
-                        latest = undefined},
+            D2 = D1#data{log   = log_msg(rcv, ?SIGNED, Msg, D1#data.log)
+                       , state = State
```
Comment (grzegorz225) [inline apps/aechannel/src/aesc_fsm.erl]: Add space

File: apps/aechannel/src/aesc_fsm.erl [hunk 10]
```diff
@@ -618,81 +634,100 @@ awaiting_signature(cast, {?SIGNED, ?FND_CREATED, SignedTx} = Msg,
...
+                   #data{op = #op_sign{ tag = ?SHUTDOWN_ACK
+                                      , data = OpData0}} = D) ->
+    #op_data{updates = Updates} = OpData0,
     maybe_check_sigs(SignedTx, Updates, aesc_close_mutual_tx, not_close_mutual_tx, both,
         fun() ->
             D1 = send_shutdown_ack_msg(SignedTx, D),
-            D2 = D1#data{latest = undefined,
-                        log    = log_msg(rcv, ?SIGNED, Msg, D1#data.log)},
+            D2 = D1#data{op = ?NO_OP
+                       , log = log_msg(rcv, ?SIGNED, Msg, D1#data.log)},
```
Comment (grzegorz225) [inline apps/aechannel/src/aesc_fsm.erl]: Add space

File: apps/aechannel/src/aesc_fsm.erl [hunk 11]
```diff
@@ -618,81 +634,100 @@ awaiting_signature(cast, {?SIGNED, ?FND_CREATED, SignedTx} = Msg,
...
         end, D);
 awaiting_signature(cast, {?SIGNED, close_solo_tx, SignedTx} = Msg,
-                   #data{latest = {sign, close_solo_tx, _CSTx, Updates}} = D) ->
-    D1 = D#data{log = log_msg(rcv, ?SIGNED, Msg, D#data.log),
-                latest = undefined},
+                   #data{op = #op_sign{ tag = close_solo_tx
+                                      , data = OpData}} = D) ->
+    #op_data{updates = Updates} = OpData,
+    D1 = D#data{log = log_msg(rcv, ?SIGNED, Msg, D#data.log)
+              , op = ?NO_OP},
```
Comment (grzegorz225) [inline apps/aechannel/src/aesc_fsm.erl]: Add space

File: apps/aechannel/src/aesc_fsm.erl [hunk 12]
```diff
@@ -2283,27 +2361,33 @@ send_conflict_msg(?WDRAW_ERR, Sn, Msg) ->
...
+                         , data = #op_data{ signed_tx = SignedTx
+                                          , block_hash = BlockHash 
+                                          , updates = Updates}}
           , log    = log_msg(req, sign, Msg, D#data.log)}.
 
 start_min_depth_watcher(Type, SignedTx, Updates,
-                        #data{watcher = Watcher0,
-                              opts = #{minimum_depth := MinDepth}} = D) ->
+                        #data{watcher = Watcher0
+                            , op = Op
```
Comment (grzegorz225) [inline apps/aechannel/src/aesc_fsm.erl]: Add space

**Pull request merged**